### PR TITLE
CURATOR-725: Allow for global compression

### PR DESCRIPTION
**Pull request by HoustonPutman**
https://issues.apache.org/jira/browse/CURATOR-725

Allow users to use a CuratorFramework setting to enable compression on all create/setData/getData requests. When using this flag with a custom `CompressionProvider`, this can easily allow for a filtered compression implementation.

### CODE REVIEW

**Review by dsmiley — APPROVED** [review]
+1 LGTM!
Disclaimer: I'm not a Curator internals expert

**Review by kezhuw — COMMENTED** [review]
+1 to global compression

> When using this flag with a custom CompressionProvider, this can easily allow for a filtered compression implementation.

I saw no per-operation way to opt out compression if it is enabled global while we have method to turn it on if it is not enabled globally. People will have to override `CompressionProvider` which complicate things.

I checked `CreateBuilder`, `SetDataBuilder`, `AsyncCreateBuilder`,  `AsyncSetDataBuilder` and `GetDataBuilder`.

 I am thinking whether we can solve this by:
1. Per operation `compression(bool compression)` to override global setting. I guess we probably could unify `Compressible` and `Decompressible`.
2. Independent global filter to separate compression filter and the algorithm. This way both parts are concentrated. And we probably could make the filter composable.

```java
public interface CompressionFilter {
    boolean compressed(String path);

    CompressionFilter ENABLED = path -> true;
    CompressionFilter DISABLED = path -> false;

    Function<Collection<CompressionFilter>, CompressionFilter> OR = filters -> path -> {
        for (CompressionFilter filter : filters) {
            if (filter.compressed(path)) {
                return true;
            }
        }
        return false;
    };

    Function<String, CompressionFilter> CHILD_NODES = parent_path -> path -> {
        ZKPaths.PathAndNode nodes = ZKPaths.getPathAndNode(path);
        return nodes.getPath().equals(parent_path);
    };

    Function<String, CompressionFilter> THIS_NODE = this_path -> path -> path.equals(this_path);

    default void example() {
        CompressionFilter this_or_child_nodes =
                OR.apply(Arrays.asList(CHILD_NODES.apply("/xx"), THIS_NODE.apply("/xx")));
    }
}
```

**Review by tisonkun — COMMENTED** [review]
Generally looks good. Thanks for your contribution!

One suggestion is that I'd prefer just name the option "compressionEnabled" and thus `enableCompression` accordingly. It's configured to the factory/framework and thus bound to that scope, instead of fully global anyway.

**Comment by HoustonPutman** [comment]
> +1 to global compression
> 
> > When using this flag with a custom CompressionProvider, this can easily allow for a filtered compression implementation.
> 
> I saw no per-operation way to opt out compression if it is enabled global while we have method to turn it on if it is not enabled globally. People will have to override `CompressionProvider` which complicate things.
> 
> I checked `CreateBuilder`, `SetDataBuilder`, `AsyncCreateBuilder`, `AsyncSetDataBuilder` and `GetDataBuilder`.

So I've gone ahead and added negative methods for both `Compressible` and `Decompressible`. I need to add tests for everything, but the Async APIs should be fully covered now too. Lots of ways to do everything, so lots of things to test here to make sure nothing got missed....

**Review by kezhuw — COMMENTED** [review]
> So I've gone ahead and added negative methods for both Compressible and Decompressible. I need to add tests for everything, but the Async APIs should be fully covered now too. Lots of ways to do everything, so lots of things to test here to make sure nothing got missed....

Thank you for the hard work! The result looks good! I left some minor inline comments.

File: curator-x-async/src/main/java/org/apache/curator/x/async/details/AsyncCreateBuilderImpl.java [hunk 1]
```diff
@@ -167,7 +167,9 @@ private AsyncStage<String> internalForPath(String path, byte[] data, boolean use
                         || options.contains(CreateOption.createParentsAsContainers),
                 options.contains(CreateOption.createParentsAsContainers),
                 options.contains(CreateOption.doProtected),
-                options.contains(CreateOption.compress),
+                options.contains(CreateOption.compress)
+                        ? true
+                        : options.contains(CreateOption.uncompress) ? false : client.compressionEnabled(),
```
[the remark this answers is not in the document]
Comment (HoustonPutman) [reply curator-x-async/src/main/java/org/apache/curator/x/async/details/AsyncCreateBuilderImpl.java:172]: I also got this from IDEA, but felt that mine was a bit easier to read?

File: curator-x-async/src/main/java/org/apache/curator/x/async/modeled/details/ModeledFrameworkImpl.java [hunk 2]
```diff
@@ -407,15 +414,27 @@ public AsyncStage<List<CuratorTransactionResult>> inTransaction(List<CuratorOp>
...
 
-    private boolean isCompressed() {
-        return modelSpec.createOptions().contains(CreateOption.compress);
+    private Boolean isCompressed() {
+        if (modelSpec.createOptions().contains(CreateOption.compress)) {
+            return Boolean.TRUE;
+        } else if (modelSpec.createOptions().contains(CreateOption.uncompress)) {
+            return Boolean.FALSE;
+        } else {
+            return null;
```
[the remark this answers is not in the document]
Comment (HoustonPutman) [reply curator-x-async/src/main/java/org/apache/curator/x/async/modeled/details/ModeledFrameworkImpl.java:423]: Ahh I didn't see the client.unwrap(). In that case, there's no need to use the `Boolean` class, we can more easily just use `boolean`. I'll change that throughout the async parts of the code. Should remove the need for the binary incompatible changes you mentioned above.

File: curator-x-async/src/main/java/org/apache/curator/x/async/api/AsyncTransactionCreateBuilder.java [hunk 3]
```diff
@@ -74,7 +82,7 @@ public interface AsyncTransactionCreateBuilder extends AsyncPathAndBytesable<Cur
      * @see #compressed()
      * @return this
      */
-    AsyncPathAndBytesable<CuratorOp> withOptions(CreateMode createMode, List<ACL> aclList, boolean compressed);
+    AsyncPathAndBytesable<CuratorOp> withOptions(CreateMode createMode, List<ACL> aclList, Boolean compressed);
```
[the remark this answers is not in the document]
Comment (HoustonPutman) [reply curator-x-async/src/main/java/org/apache/curator/x/async/api/AsyncTransactionCreateBuilder.java:85]: > I am thinking it might be more composable for `AsyncXzyBuilder` to return itself for methods `withAbc`. But it is not binary compatible.

Yes, I absolutely wanted to do that, but thought it would create back-compat concerns...

I do like your suggestion about the new method that we could then compress or uncompressed.

**Review by tisonkun — APPROVED** [review]
LGTM. Thanks for your contribution!

File: curator-framework/src/test/java/org/apache/curator/framework/imps/TestCompressionInTransactionOld.java [hunk 4]
```diff
@@ -172,4 +172,59 @@ public void testCreateCompressedAndUncompressed() throws Exception {
...
+            assertArrayEquals(data1, client.getData().decompressed().forPath(path1));
+            assertArrayEquals(data2, client.getData().decompressed().forPath(path2));
+
+            // Set the nodes, path1 compressed, path2 uncompressed.
+            client.inTransaction()
+                    .setData()
+                    .forPath(path1, data2)
+                    .and()
+                    .setData()
+                    .forPath(path2, data1)
```
[the remark this answers is not in the document]
Comment (kezhuw) [reply curator-framework/src/test/java/org/apache/curator/framework/imps/TestCompressionInTransactionOld.java:216]: The code and comment diverge here.

**Review by kezhuw — APPROVED** [review]
There is a mismatch between code and comment in test, I left comment there.

> I need to add tests for everything, but the Async APIs should be fully covered now too. Lots of ways to do everything, so lots of things to test here to make sure nothing got missed....

I expected there are test updates.

**Comment by HoustonPutman** [comment]
Ok, should have fully added tests here. Not every single method is tested, but every use case is with lots of corner cases.

I think this should be ready to go

**Comment by tisonkun** [comment]
@HoustonPutman Thanks for your contribution! Merged now :D
